File: vscode-sqltools/src/test/suite/schema-cache.test.ts
import * as assert from 'assert';
import { SchemaCache } from '../../ls/schema-cache';

suite('SchemaCache - Buckets', () => {
  test('cache hit within TTL: fetchFn is not called a second time', async () => {
    const cache = new SchemaCache(1000);
    let callCount = 0;
    const fetchFn = async () => {
      callCount++;
      return [{ id: 'in.c-main' }, { id: 'out.c-results' }];
    };

    const first = await cache.getBuckets(fetchFn);
    const second = await cache.getBuckets(fetchFn);

    assert.strictEqual(callCount, 1, 'fetchFn should be called only once within TTL');
    assert.deepStrictEqual(first, second);
    assert.strictEqual(first.length, 2);
  });

  test('cache miss after TTL expiry: fetchFn is called again', async () => {
    const cache = new SchemaCache(10); // 10ms TTL
    let callCount = 0;
    const fetchFn = async () => {
      callCount++;
      return [{ id: `bucket-${callCount}` }];
    };

    await cache.getBuckets(fetchFn);
    assert.strictEqual(callCount, 1);

    // Wait for TTL to expire
    await new Promise((resolve) => setTimeout(resolve, 20));

    const result = await cache.getBuckets(fetchFn);
    assert.strictEqual(callCount, 2, 'fetchFn should be called again after TTL expiry');
    assert.deepStrictEqual(result, [{ id: 'bucket-2' }]);
  });

  test('stale-on-error: returns old cached data when fetch fails', async () => {
    const cache = new SchemaCache(10); // 10ms TTL
    const originalData = [{ id: 'in.c-main' }];
    let shouldFail = false;

    const fetchFn = async () => {
      if (shouldFail) {
        throw new Error('API unavailable');
      }
      return originalData;
    };

    // First call succeeds and populates cache
    const first = await cache.getBuckets(fetchFn);
    assert.deepStrictEqual(first, originalData);

    // Wait for TTL to expire
    await new Promise((resolve) => setTimeout(resolve, 20));

    // Second call fails but returns stale data
    shouldFail = true;
    const second = await cache.getBuckets(fetchFn);
    assert.deepStrictEqual(second, originalData, 'Should return stale data on fetch error');
  });

  test('no stale data available: throws error on first call failure', async () => {
    const cache = new SchemaCache(1000);
    const fetchFn = async () => {
      throw new Error('API unavailable');
    };

    await assert.rejects(
      () => cache.getBuckets(fetchFn),
      /API unavailable/,
      'Should throw when no stale data is available'
    );
  });
});

suite('SchemaCache - Tables', () => {
  test('cache hit within TTL: fetchFn is not called a second time', async () => {
    const cache = new SchemaCache(1000);
    let callCount = 0;
    const fetchFn = async () => {
      callCount++;
      return [{ name: 'users' }, { name: 'orders' }];
    };

    const first = await cache.getTables('in.c-main', fetchFn);
    const second = await cache.getTables('in.c-main', fetchFn);

    assert.strictEqual(callCount, 1, 'fetchFn should be called only once within TTL');
    assert.deepStrictEqual(first, second);
    assert.strictEqual(first.length, 2);
  });

  test('cache miss after TTL expiry: fetchFn is called again', async () => {
    const cache = new SchemaCache(10); // 10ms TTL
    let callCount = 0;
    const fetchFn = async () => {
      callCount++;
      return [{ name: `table-${callCount}` }];
    };

    await cache.getTables('in.c-main', fetchFn);
    assert.strictEqual(callCount, 1);

    // Wait for TTL to expire
    await new Promise((resolve) => setTimeout(resolve, 20));

    const result = await cache.getTables('in.c-main', fetchFn);
    assert.strictEqual(callCount, 2, 'fetchFn should be called again after TTL expiry');
    assert.deepStrictEqual(result, [{ name: 'table-2' }]);
  });

  test('stale-on-error: returns old cached data when fetch fails', async () => {
    const cache = new SchemaCache(10); // 10ms TTL
    const originalData = [{ name: 'users' }];
    let shouldFail = false;

    const fetchFn = async () => {
      if (shouldFail) {
        throw new Error('API unavailable');
      }
      return originalData;
    };

    // First call succeeds and populates cache
    const first = await cache.getTables('in.c-main', fetchFn);
    assert.deepStrictEqual(first, originalData);

    // Wait for TTL to expire
    await new Promise((resolve) => setTimeout(resolve, 20));

    // Second call fails but returns stale data
    shouldFail = true;
    const second = await cache.getTables('in.c-main', fetchFn);
    assert.deepStrictEqual(second, originalData, 'Should return stale data on fetch error');
  });

  test('no stale data available: throws error on first call failure', async () => {
    const cache = new SchemaCache(1000);
    const fetchFn = async () => {
      throw new Error('API unavailable');
    };

    await assert.rejects(
      () => cache.getTables('in.c-main', fetchFn),
      /API unavailable/,
      'Should throw when no stale data is available'
    );
  });

  test('per-bucket caching: different buckets have separate caches', async () => {
    const cache = new SchemaCache(1000);
    let callCount = 0;

    const fetchMain = async () => {
      callCount++;
      return [{ name: 'users' }];
    };
    const fetchOutput = async () => {
      callCount++;
      return [{ name: 'results' }];
    };

    const mainTables = await cache.getTables('in.c-main', fetchMain);
    const outputTables = await cache.getTables('out.c-output', fetchOutput);

    assert.strictEqual(callCount, 2, 'Each bucket should trigger its own fetch');
    assert.deepStrictEqual(mainTables, [{ name: 'users' }]);
    assert.deepStrictEqual(outputTables, [{ name: 'results' }]);

    // Second call to in.c-main should use cache
    await cache.getTables('in.c-main', fetchMain);
    assert.strictEqual(callCount, 2, 'Cached bucket should not trigger another fetch');
  });
});

suite('SchemaCache - Invalidation', () => {
  test('invalidateBucket clears only the specified bucket', async () => {
    const cache = new SchemaCache(1000);
    let callCount = 0;

    const fetchFn = async () => {
      callCount++;
      return [{ name: `table-${callCount}` }];
    };

    await cache.getTables('in.c-main', fetchFn);
    await cache.getTables('out.c-output', fetchFn);
    assert.strictEqual(callCount, 2);

    // Invalidate only in.c-main
    cache.invalidateBucket('in.c-main');

    // in.c-main should fetch again
    await cache.getTables('in.c-main', fetchFn);
    assert.strictEqual(callCount, 3, 'Invalidated bucket should trigger a new fetch');

    // out.c-output should still be cached
    await cache.getTables('out.c-output', fetchFn);
    assert.strictEqual(callCount, 3, 'Non-invalidated bucket should still use cache');
  });

  test('invalidate clears all cached data', async () => {
    const cache = new SchemaCache(1000);
    let bucketCallCount = 0;
    let tableCallCount = 0;

    const fetchBuckets = async () => {
      bucketCallCount++;
      return [{ id: 'in.c-main' }];
    };
    const fetchTables = async () => {
      tableCallCount++;
      return [{ name: 'users' }];
    };

    await cache.getBuckets(fetchBuckets);
    await cache.getTables('in.c-main', fetchTables);
    assert.strictEqual(bucketCallCount, 1);
    assert.strictEqual(tableCallCount, 1);

    // Invalidate all
    cache.invalidate();

    // Both should fetch again
    await cache.getBuckets(fetchBuckets);
    await cache.getTables('in.c-main', fetchTables);
    assert.strictEqual(bucketCallCount, 2, 'Buckets should be re-fetched after full invalidation');
    assert.strictEqual(tableCallCount, 2, 'Tables should be re-fetched after full invalidation');
  });

  test('invalidate on empty cache does not throw', () => {
    const cache = new SchemaCache(1000);
    assert.doesNotThrow(() => cache.invalidate());
  });

  test('invalidateBucket on non-existent bucket does not throw', () => {
    const cache = new SchemaCache(1000);
    assert.doesNotThrow(() => cache.invalidateBucket('nonexistent'));
  });
});

File: jdbc-driver/src/test/java/com/keboola/jdbc/meta/SchemaCacheTest.java
package com.keboola.jdbc.meta;

import com.keboola.jdbc.http.StorageApiClient;
import com.keboola.jdbc.http.model.Bucket;
import com.keboola.jdbc.http.model.TableInfo;
import com.keboola.jdbc.exception.KeboolaJdbcException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

/**
 * Unit tests for SchemaCache - verifies caching behaviour, TTL expiry, and cache invalidation.
 */
@ExtendWith(MockitoExtension.class)
class SchemaCacheTest {

    @Mock
    private StorageApiClient mockApiClient;

    private SchemaCache schemaCache;

    // A very short TTL (1 ms) used in tests that need to simulate TTL expiry
    private static final long SHORT_TTL_MS = 1L;
    // A very long TTL (10 minutes) used in tests that need a valid cache
    private static final long LONG_TTL_MS = 600_000L;

    @BeforeEach
    void setUp() {
        // Default cache with long TTL - individual tests can create their own with short TTL
        schemaCache = new SchemaCache(mockApiClient, LONG_TTL_MS);
    }

    // -------------------------------------------------------------------------
    // getBuckets() - API call behaviour
    // -------------------------------------------------------------------------

    @Test
    void getBuckets_firstCall_callsApi() throws KeboolaJdbcException {
        List<Bucket> expected = Arrays.asList(
                new Bucket("in.c-main", "main", "in", "", "snowflake"),
                new Bucket("out.c-reports", "reports", "out", "", "snowflake")
        );
        when(mockApiClient.listBuckets()).thenReturn(expected);

        List<Bucket> result = schemaCache.getBuckets();

        assertEquals(2, result.size());
        verify(mockApiClient, times(1)).listBuckets();
    }

    @Test
    void getBuckets_secondCallWithinTtl_usesCacheAndDoesNotCallApiAgain() throws KeboolaJdbcException {
        List<Bucket> expected = Collections.singletonList(
                new Bucket("in.c-main", "main", "in", "", "snowflake")
        );
        when(mockApiClient.listBuckets()).thenReturn(expected);

        // First call - populates cache
        schemaCache.getBuckets();
        // Second call - should use cache
        List<Bucket> result = schemaCache.getBuckets();

        assertEquals(1, result.size());
        // API should only be called once despite two getBuckets() invocations
        verify(mockApiClient, times(1)).listBuckets();
    }

    @Test
    void getBuckets_afterTtlExpiry_refreshesCacheByCallingApiAgain() throws KeboolaJdbcException, InterruptedException {
        SchemaCache shortTtlCache = new SchemaCache(mockApiClient, SHORT_TTL_MS);

        List<Bucket> firstBatch = Collections.singletonList(
                new Bucket("in.c-first", "first", "in", "", "snowflake")
        );
        List<Bucket> secondBatch = Arrays.asList(
                new Bucket("in.c-first", "first", "in", "", "snowflake"),
                new Bucket("in.c-second", "second", "in", "", "snowflake")
        );

        when(mockApiClient.listBuckets())
                .thenReturn(firstBatch)
                .thenReturn(secondBatch);

        // First call
        List<Bucket> first = shortTtlCache.getBuckets();
        assertEquals(1, first.size());

        // Wait for TTL to expire
        Thread.sleep(SHORT_TTL_MS + 5);

        // Second call should refresh
        List<Bucket> second = shortTtlCache.getBuckets();
        assertEquals(2, second.size());

        verify(mockApiClient, times(2)).listBuckets();
    }

    @Test
    void getBuckets_returnsEmptyList_whenApiReturnsEmptyList() throws KeboolaJdbcException {
        when(mockApiClient.listBuckets()).thenReturn(Collections.emptyList());

        List<Bucket> result = schemaCache.getBuckets();

        assertNotNull(result);
        assertTrue(result.isEmpty());
    }

    @Test
    void getBuckets_whenApiThrowsAndNoCachedData_returnsEmptyList() throws KeboolaJdbcException {
        when(mockApiClient.listBuckets()).thenThrow(new KeboolaJdbcException("Network error"));

        List<Bucket> result = schemaCache.getBuckets();

        assertNotNull(result);
        assertTrue(result.isEmpty(), "Should return empty list when API fails and no cached data exists");
    }

    @Test
    void getBuckets_whenApiThrowsButCachedDataExists_returnsStaleCachedData() throws KeboolaJdbcException, InterruptedException {
        SchemaCache shortTtlCache = new SchemaCache(mockApiClient, SHORT_TTL_MS);

        List<Bucket> staleData = Collections.singletonList(
                new Bucket("in.c-stale", "stale", "in", "", "snowflake")
        );

        when(mockApiClient.listBuckets())
                .thenReturn(staleData)
                .thenThrow(new KeboolaJdbcException("API unavailable"));

        // Populate cache
        shortTtlCache.getBuckets();

        // Wait for TTL to expire
        Thread.sleep(SHORT_TTL_MS + 5);

        // Second call - API fails but stale data should be returned
        List<Bucket> result = shortTtlCache.getBuckets();

        assertEquals(1, result.size());
        assertEquals("in.c-stale", result.get(0).getId());
    }

    // -------------------------------------------------------------------------
    // getTables() - API call behaviour
    // -------------------------------------------------------------------------

    @Test
    void getTables_firstCall_callsApiWithCorrectBucketId() throws KeboolaJdbcException {
        String bucketId = "in.c-main";
        List<TableInfo> tables = Collections.singletonList(
                new TableInfo("in.c-main.orders", "orders", null, Collections.emptyList(), Collections.emptyMap())
        );
        when(mockApiClient.listTables(bucketId)).thenReturn(tables);

        List<TableInfo> result = schemaCache.getTables(bucketId);

        assertEquals(1, result.size());
        assertEquals("in.c-main.orders", result.get(0).getId());
        verify(mockApiClient, times(1)).listTables(bucketId);
    }

    @Test
    void getTables_secondCallWithinTtl_usesCacheForSameBucket() throws KeboolaJdbcException {
        String bucketId = "in.c-main";
        List<TableInfo> tables = Collections.singletonList(
                new TableInfo("in.c-main.orders", "orders", null, Collections.emptyList(), Collections.emptyMap())
        );
        when(mockApiClient.listTables(bucketId)).thenReturn(tables);

        // Populate cache
        schemaCache.getTables(bucketId);
        // Should use cache
        List<TableInfo> result = schemaCache.getTables(bucketId);

        assertEquals(1, result.size());
        verify(mockApiClient, times(1)).listTables(bucketId);
    }

    @Test
    void getTables_differentBuckets_callsApiForEach() throws KeboolaJdbcException {
        String bucket1 = "in.c-main";
        String bucket2 = "out.c-reports";

        when(mockApiClient.listTables(bucket1)).thenReturn(Collections.emptyList());
        when(mockApiClient.listTables(bucket2)).thenReturn(Collections.emptyList());

        schemaCache.getTables(bucket1);
        schemaCache.getTables(bucket2);

        // Each bucket must be fetched independently
        verify(mockApiClient, times(1)).listTables(bucket1);
        verify(mockApiClient, times(1)).listTables(bucket2);
    }

    @Test
    void getTables_afterTtlExpiry_refreshesCacheForBucket() throws KeboolaJdbcException, InterruptedException {
        SchemaCache shortTtlCache = new SchemaCache(mockApiClient, SHORT_TTL_MS);
        String bucketId = "in.c-main";

        when(mockApiClient.listTables(bucketId)).thenReturn(Collections.emptyList());

        shortTtlCache.getTables(bucketId);

        Thread.sleep(SHORT_TTL_MS + 5);

        shortTtlCache.getTables(bucketId);

        verify(mockApiClient, times(2)).listTables(bucketId);
    }

    // -------------------------------------------------------------------------
    // invalidate()
    // -------------------------------------------------------------------------

    @Test
    void invalidate_clearsAllCaches_causesApiCallOnNextAccess() throws KeboolaJdbcException {
        String bucketId = "in.c-main";

        when(mockApiClient.listBuckets()).thenReturn(Collections.emptyList());
        when(mockApiClient.listTables(bucketId)).thenReturn(Collections.emptyList());

        // Populate both caches
        schemaCache.getBuckets();
        schemaCache.getTables(bucketId);

        // Invalidate everything
        schemaCache.invalidate();

        // Access again - should call API
        schemaCache.getBuckets();
        schemaCache.getTables(bucketId);

        verify(mockApiClient, times(2)).listBuckets();
        verify(mockApiClient, times(2)).listTables(bucketId);
    }

    @Test
    void invalidate_calledMultipleTimes_doesNotThrow() throws KeboolaJdbcException {
        when(mockApiClient.listBuckets()).thenReturn(Collections.emptyList());
        schemaCache.getBuckets();

        assertDoesNotThrow(() -> {
            schemaCache.invalidate();
            schemaCache.invalidate();
        });
    }

    @Test
    void invalidate_afterCachingBuckets_nextGetBucketsCallsApi() throws KeboolaJdbcException {
        when(mockApiClient.listBuckets()).thenReturn(Collections.emptyList());

        schemaCache.getBuckets(); // First call - populates cache
        schemaCache.invalidate(); // Clear cache
        schemaCache.getBuckets(); // Should call API again

        verify(mockApiClient, times(2)).listBuckets();
    }
}
